### PR TITLE
Add Classy versions of Era witness functions

### PR DESCRIPTION
**Pull request by locallycompact**
# Changelog

```yaml
- description: |
    Add `IsAllegraBasedEra`, `IsAlonzoBasedEra`, `IsBabbageBasedEra`, `IsConwayBasedEra`, `IsMaryBasedEra` type classes.
    Add `ToAlonzoScript` and `HasScriptLanguageInEra` type classes.
# uncomment types applicable to the change:
  type:
  - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - refactoring    # QoL changes
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Additional context for the PR goes here. If the PR fixes a particular issue please provide a [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=) to the issue.

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by locallycompact** [comment]
Any opinions on this @newhoggy ?

**Comment by smelc** [comment]
cc @carbolymer

File: cardano-api/internal/Cardano/Api/Class/ToAlonzoScript.hs
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Class/ToAlonzoScript.hs:1]: I think this should go to `Cardano.Api.Script`

**Review by carbolymer — COMMENTED** [review]
Looks good, few regards about naming things

**Comment by palas** [comment]
FYI: I have rebased your branch because we have done changes to the formatting. I have made a copy of the unrebased branch here: [backup/locallycompact/classy](https://github.com/IntersectMBO/cardano-api/tree/backup/locallycompact/classy)

**Comment by locallycompact** [comment]
@carbolymer Could I get a re-review on this?

**Review by smelc — APPROVED** [review]

**Review by carbolymer — APPROVED** [review]
LGTM! Thanks!

**Comment by smelc** [comment]
Closing as it was duplicated in https://github.com/IntersectMBO/cardano-api/pull/585 (which I can sign, since it's on this repo; instead of a clone).

**Pull request closed without merging**